networking/nat/src/main/java/tech/pegasys/teku/networking/nat/NatManager.java

```diff
@@ -0,0 +1,419 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.networking.nat;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static tech.pegasys.teku.infrastructure.async.FutureUtil.ignoreFuture;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jupnp.UpnpService;
+import org.jupnp.UpnpServiceImpl;
+import org.jupnp.model.action.ActionInvocation;
+import org.jupnp.model.message.UpnpResponse;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.model.types.UnsignedIntegerFourBytes;
+import org.jupnp.model.types.UnsignedIntegerTwoBytes;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.jupnp.support.igd.callback.GetExternalIP;
+import org.jupnp.support.igd.callback.PortMappingAdd;
+import org.jupnp.support.igd.callback.PortMappingDelete;
+import org.jupnp.support.model.PortMapping;
+import tech.pegasys.teku.infrastructure.async.FutureUtil;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.service.serviceutils.Service;
+
+@SuppressWarnings("unchecked")
+public class NatManager extends Service {
+  protected static final Logger LOG = LogManager.getLogger();
+
+  static final String SERVICE_TYPE_WAN_IP_CONNECTION = "WANIPConnection";
+
+  private final UpnpService upnpService;
+  private final RegistryListener registryListener;
+  private final SafeFuture<String> externalIpQueryFuture = new SafeFuture<>();
+
+  private final Map<String, SafeFuture<RemoteService>> recognizedServices = new HashMap<>();
+  private volatile Queue<NatPortMapping> forwardedPorts = new ConcurrentLinkedQueue<>();
```

Comment:
There's no need for this to be `volatile` as it's never reassigned.  Should just be final.
```suggestion
  private final Queue<NatPortMapping> forwardedPorts = new ConcurrentLinkedQueue<>();
```